**Código da aula 3:**

In [ ]:
import pandas as pd

# Carregar os dados
data_path = '/content/student-mat.csv'
data = pd.read_csv(data_path)

# Exibir as primeiras linhas para entender a estrutura dos dados
data.head()


school;sex;age;address;famsize;Pstatus;Medu;Fedu;Mjob;Fjob;reason;guardian;traveltime;studytime;failures;schoolsup;famsup;paid;activities;nursery;higher;internet;romantic;famrel;freetime;goout;Dalc;Walc;health;absences;G1;G2;G3
0  GP;"F";18;"U";"GT3";"A";4;4;"at_home";"teacher...                                                                                                                                                                                 
1  GP;"F";17;"U";"GT3";"T";1;1;"at_home";"other";...                                                                                                                                                                                 
2  GP;"F";15;"U";"LE3";"T";1;1;"at_home";"other";...                                                                                                                                                                                 
3  GP;"F";15;"U";"GT3";"T";4;2;"health";"services...                                                                                                                                                                                 
4  GP;"F";16;"U";"GT3";"T";3;3;"other";"other";"h...

In [ ]:
# Corrigir o separador e recarregar os dados
data = pd.read_csv(data_path, sep=';')

# Exibir novamente as primeiras linhas
data.head()


school sex  age address famsize Pstatus  Medu  Fedu     Mjob      Fjob  ...  \
0     GP   F   18       U     GT3       A     4     4  at_home   teacher  ...   
1     GP   F   17       U     GT3       T     1     1  at_home     other  ...   
2     GP   F   15       U     LE3       T     1     1  at_home     other  ...   
3     GP   F   15       U     GT3       T     4     2   health  services  ...   
4     GP   F   16       U     GT3       T     3     3    other     other  ...   

  famrel freetime  goout  Dalc  Walc health absences  G1  G2  G3  
0      4        3      4     1     1      3        6   5   6   6  
1      5        3      3     1     1      3        4   5   5   6  
2      4        3      2     2     3      3       10   7   8  10  
3      3        2      2     1     1      5        2  15  14  15  
4      4        3      2     1     2      5        4   6  10  10  

[5 rows x 33 columns]

In [ ]:
from sklearn.model_selection import train_test_split

# Definir a variável alvo (target)
# Consideraremos 'sucesso' como ter uma nota final (G3) igual ou superior a 10
data['success'] = data['G3'] >= 10

# Selecionar variáveis preditoras de interesse
# Incluindo informações básicas e indicadores de desempenho e hábitos de estudo
predictors = ['sex', 'age', 'studytime', 'failures', 'absences']

# Converter variáveis categóricas em dummy variables
X = pd.get_dummies(data[predictors], drop_first=True)

# Definir a variável alvo
y = data['success']

# Dividir os dados em conjunto de treinamento e teste
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Exibir as dimensões dos conjuntos de treinamento e teste para verificar
X_train.shape, X_test.shape, y_train.shape, y_test.shape


((316, 5), (79, 5), (316,), (79,))

In [ ]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, classification_report

# Instanciar o modelo de Regressão Logística
model = LogisticRegression(max_iter=1000)

# Treinar o modelo com o conjunto de treinamento
model.fit(X_train, y_train)

# Prever os resultados para o conjunto de teste
y_pred = model.predict(X_test)

# Calcular a acurácia e exibir o relatório de classificação
accuracy = accuracy_score(y_test, y_pred)
report = classification_report(y_test, y_pred)

accuracy, report


(0.7341772151898734,
 '              precision    recall  f1-score   support\n\n       False       0.75      0.33      0.46        27\n        True       0.73      0.94      0.82        52\n\n    accuracy                           0.73        79\n   macro avg       0.74      0.64      0.64        79\nweighted avg       0.74      0.73      0.70        79\n')

In [ ]:
# Verificar os tipos de dados das colunas
data_types = data.dtypes

# Ver o formato do DataFrame
data_shape = data.shape

data_types, data_shape


(school        object
 sex           object
 age            int64
 address       object
 famsize       object
 Pstatus       object
 Medu           int64
 Fedu           int64
 Mjob          object
 Fjob          object
 reason        object
 guardian      object
 traveltime     int64
 studytime      int64
 failures       int64
 schoolsup     object
 famsup        object
 paid          object
 activities    object
 nursery       object
 higher        object
 internet      object
 romantic      object
 famrel         int64
 freetime       int64
 goout          int64
 Dalc           int64
 Walc           int64
 health         int64
 absences       int64
 G1             int64
 G2             int64
 G3             int64
 success         bool
 dtype: object,
 (395, 34))

In [ ]:
# Configurar a precisão dos números para duas casas decimais
pd.set_option('display.precision', 2)

# Visualizar uma descrição estatística das variáveis numéricas
description = data.describe()

description


age    Medu    Fedu  traveltime  studytime  failures  famrel  \
count  395.00  395.00  395.00      395.00     395.00    395.00  395.00   
mean    16.70    2.75    2.52        1.45       2.04      0.33    3.94   
std      1.28    1.09    1.09        0.70       0.84      0.74    0.90   
min     15.00    0.00    0.00        1.00       1.00      0.00    1.00   
25%     16.00    2.00    2.00        1.00       1.00      0.00    4.00   
50%     17.00    3.00    2.00        1.00       2.00      0.00    4.00   
75%     18.00    4.00    3.00        2.00       2.00      0.00    5.00   
max     22.00    4.00    4.00        4.00       4.00      3.00    5.00   

       freetime   goout    Dalc    Walc  health  absences      G1      G2  \
count    395.00  395.00  395.00  395.00  395.00    395.00  395.00  395.00   
mean       3.24    3.11    1.48    2.29    3.55      5.71   10.91   10.71   
std        1.00    1.11    0.89    1.29    1.39      8.00    3.32    3.76   
min        1.00    1.00    1.00    1.00    1.00      0.00    3.00    0.00   
25%        3.00    2.00    1.00    1.00    3.00      0.00    8.00    9.00   
50%        3.00    3.00    1.00    2.00    4.00      4.00   11.00   11.00   
75%        4.00    4.00    2.00    3.00    5.00      8.00   13.00   13.00   
max        5.00    5.00    5.00    5.00    5.00     75.00   19.00   19.00   

           G3  
count  395.00  
mean    10.42  
std      4.58  
min      0.00  
25%      8.00  
50%     11.00  
75%     14.00  
max     20.00

In [ ]:
# Verificar a presença de valores faltantes no DataFrame
missing_values = data.isnull().sum()

missing_values


school        0
sex           0
age           0
address       0
famsize       0
Pstatus       0
Medu          0
Fedu          0
Mjob          0
Fjob          0
reason        0
guardian      0
traveltime    0
studytime     0
failures      0
schoolsup     0
famsup        0
paid          0
activities    0
nursery       0
higher        0
internet      0
romantic      0
famrel        0
freetime      0
goout         0
Dalc          0
Walc          0
health        0
absences      0
G1            0
G2            0
G3            0
success       0
dtype: int64

In [ ]:
# Verificar a distribuição das classes na variável alvo 'success'
class_counts = data.groupby('success').size()

class_counts


success
False    130
True     265
dtype: int64

In [ ]:
# Remover instâncias com dados faltantes (nesse caso, não alterará o DataFrame)
data_cleaned = data.dropna()

# Verificar novamente as dimensões para confirmar que nenhum dado foi removido
data_cleaned_shape = data_cleaned.shape

data_cleaned_shape


(395, 34)

In [ ]:
# Remover instâncias duplicadas
data_no_duplicates = data_cleaned.drop_duplicates()

# Verificar as dimensões do DataFrame após remover duplicatas
data_no_duplicates_shape = data_no_duplicates.shape

data_no_duplicates_shape


(395, 34)

In [ ]:
# Realizar o teste de Shapiro-Wilk diretamente no DataFrame completo para as variáveis 'age' e 'absences'
shapiro_results_complete = {var: stats.shapiro(data[var]) for var in ['age', 'absences']}

shapiro_results_complete


{'age': ShapiroResult(statistic=0.9105901122093201, pvalue=1.587754619892124e-14),
 'absences': ShapiroResult(statistic=0.6668332815170288, pvalue=4.140299483197548e-27)}

In [ ]:
from sklearn.preprocessing import StandardScaler

# Instanciar o objeto StandardScaler
scaler = StandardScaler()

# Ajustar o scaler apenas com os dados de treinamento
scaler.fit(X_train[['age', 'absences']])

# Transformar os dados de treinamento e teste
X_train_scaled = X_train.copy()
X_test_scaled = X_test.copy()

X_train_scaled[['age', 'absences']] = scaler.transform(X_train[['age', 'absences']])
X_test_scaled[['age', 'absences']] = scaler.transform(X_test[['age', 'absences']])

# Exibir as primeiras linhas dos dados de treinamento escalonados para verificar
X_train_scaled.head()


age  studytime  failures  absences  sex_M
181 -0.59          2         0     -0.46      1
194 -0.59          1         0     -0.70      1
173 -0.59          2         3     -0.70      0
63  -0.59          3         0     -0.46      0
253 -0.59          1         0     -0.70      1

Sem Re-escalar os Dados: Acurácia de aproximadamente 73.4%

Apenas Padronizando: Acurácia de aproximadamente 73.4%

Apenas Normalizando: Acurácia de aproximadamente 73.4%

In [ ]:
from sklearn.preprocessing import MinMaxScaler

# Instanciar o objeto MinMaxScaler
min_max_scaler = MinMaxScaler()

# Ajustar o scaler apenas com os dados de treinamento escalonados
min_max_scaler.fit(X_train_scaled)

# Transformar os dados de treinamento e teste escalonados
X_train_min_max_scaled = min_max_scaler.transform(X_train_scaled)
X_test_min_max_scaled = min_max_scaler.transform(X_test_scaled)

# Converter as arrays transformadas de volta para DataFrame para facilitar a visualização
X_train_min_max_scaled_df = pd.DataFrame(X_train_min_max_scaled, columns=X_train_scaled.columns)
X_test_min_max_scaled_df = pd.DataFrame(X_test_min_max_scaled, columns=X_test_scaled.columns)

# Exibir as primeiras linhas dos dados de treinamento normalizados para verificar
X_train_min_max_scaled_df.head()


age  studytime  failures  absences  sex_M
0  0.14       0.33       0.0      0.03    1.0
1  0.14       0.00       0.0      0.00    1.0
2  0.14       0.33       1.0      0.00    0.0
3  0.14       0.67       0.0      0.03    0.0
4  0.14       0.00       0.0      0.00    1.0

In [ ]:
from sklearn.metrics import accuracy_score

def train_and_evaluate_model(X_train, X_test, y_train, y_test):
    """
    Treina um modelo de Regressão Logística com os dados de treinamento fornecidos e avalia a acurácia no conjunto de teste.
    """
    model = LogisticRegression(max_iter=1000)
    model.fit(X_train, y_train)
    y_pred = model.predict(X_test)
    accuracy = accuracy_score(y_test, y_pred)
    return accuracy

# Acumular resultados em um dicionário para comparação
results = {}

# 1. Sem Re-escalar os Dados
# Nota: As variáveis já foram transformadas em dummies anteriormente, então podemos reutilizar `X_train` e `X_test` originais.
results['Sem Re-escalar'] = train_and_evaluate_model(X_train, X_test, y_train, y_test)

# 2. Apenas Padronizando
# Usar `X_train_scaled` e `X_test_scaled` que foram padronizados anteriormente.
results['Apenas Padronizando'] = train_and_evaluate_model(X_train_scaled, X_test_scaled, y_train, y_test)

# 3. Apenas Normalizando
# Usar `X_train_min_max_scaled_df` e `X_test_min_max_scaled_df` que foram normalizados usando Min-Max após padronização.
results['Apenas Normalizando'] = train_and_evaluate_model(X_train_min_max_scaled_df, X_test_min_max_scaled_df, y_train, y_test)

results


{'Sem Re-escalar': 0.7341772151898734,
 'Apenas Padronizando': 0.7341772151898734,
 'Apenas Normalizando': 0.7341772151898734}

In [ ]:
from sklearn.metrics import precision_score, recall_score, f1_score, confusion_matrix

def evaluate_model_metrics(X_train, X_test, y_train, y_test):
    """
    Treina um modelo de Regressão Logística e avalia várias métricas de performance no conjunto de teste.
    """
    model = LogisticRegression(max_iter=1000)
    model.fit(X_train, y_train)
    y_pred = model.predict(X_test)

    # Calcular métricas
    accuracy = accuracy_score(y_test, y_pred)
    precision = precision_score(y_test, y_pred)
    recall = recall_score(y_test, y_pred)
    f1 = f1_score(y_test, y_pred)
    confusion = confusion_matrix(y_test, y_pred)

    return accuracy, precision, recall, f1, confusion

# Avaliar modelos para os diferentes cenários de pré-processamento
metrics_results = {}

# Sem Re-escalar os Dados
metrics_results['Sem Re-escalar'] = evaluate_model_metrics(X_train, X_test, y_train, y_test)

# Apenas Padronizando
metrics_results['Apenas Padronizando'] = evaluate_model_metrics(X_train_scaled, X_test_scaled, y_train, y_test)

# Apenas Normalizando
metrics_results['Apenas Normalizando'] = evaluate_model_metrics(X_train_min_max_scaled_df, X_test_min_max_scaled_df, y_train, y_test)

# Exibir as métricas calculadas para cada cenário
for scenario, metrics in metrics_results.items():
    print(f"{scenario}:")
    print(f"Acurácia: {metrics[0]:.4f}, Precisão: {metrics[1]:.4f}, Recall: {metrics[2]:.4f}, F1-Score: {metrics[3]:.4f}")
    print(f"Matriz de Confusão:\n{metrics[4]}")
    print("\n")


Sem Re-escalar:
Acurácia: 0.7342, Precisão: 0.7313, Recall: 0.9423, F1-Score: 0.8235
Matriz de Confusão:
[[ 9 18]
 [ 3 49]]


Apenas Padronizando:
Acurácia: 0.7342, Precisão: 0.7313, Recall: 0.9423, F1-Score: 0.8235
Matriz de Confusão:
[[ 9 18]
 [ 3 49]]


Apenas Normalizando:
Acurácia: 0.7342, Precisão: 0.7246, Recall: 0.9615, F1-Score: 0.8264
Matriz de Confusão:
[[ 8 19]
 [ 2 50]]




------------------------------------------------------------------------------------



**Código continuação da Aula 4:**

Regressão Logística

In [ ]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

def train_and_evaluate_logistic_regression(X_train, X_test, y_train, y_test):
    # Criar e treinar o modelo de Regressão Logística
    model = LogisticRegression(max_iter=1000)
    model.fit(X_train, y_train)

    # Fazer previsões com o conjunto de teste
    y_pred = model.predict(X_test)

    # Calcular e exibir a acurácia
    accuracy = accuracy_score(y_test, y_pred)
    print(f"Acurácia da Regressão Logística: {accuracy:.4f}")

# Chamar a função usando os dados escalonados
train_and_evaluate_logistic_regression(X_train_min_max_scaled_df, X_test_min_max_scaled_df, y_train, y_test)


Acurácia da Regressão Logística: 0.7342


In [ ]:
from sklearn.tree import DecisionTreeClassifier

def train_and_evaluate_decision_tree(X_train, X_test, y_train, y_test):
    # Criar e treinar o modelo de Árvore de Decisão
    model = DecisionTreeClassifier()
    model.fit(X_train, y_train)

    # Fazer previsões com o conjunto de teste
    y_pred = model.predict(X_test)

    # Calcular e exibir a acurácia
    accuracy = accuracy_score(y_test, y_pred)
    print(f"Acurácia da Árvore de Decisão: {accuracy:.4f}")

# Chamar a função usando os dados escalonados
train_and_evaluate_decision_tree(X_train_min_max_scaled_df, X_test_min_max_scaled_df, y_train, y_test)


Acurácia da Árvore de Decisão: 0.6203


In [ ]:
from sklearn.svm import SVC

def train_and_evaluate_svm(X_train, X_test, y_train, y_test):
    # Criar e treinar o modelo SVM
    model = SVC()
    model.fit(X_train, y_train)

    # Fazer previsões com o conjunto de teste
    y_pred = model.predict(X_test)

    # Calcular e exibir a acurácia
    accuracy = accuracy_score(y_test, y_pred)
    print(f"Acurácia do SVM: {accuracy:.4f}")

# Chamar a função usando os dados escalonados
train_and_evaluate_svm(X_train_min_max_scaled_df, X_test_min_max_scaled_df, y_train, y_test)


Acurácia do SVM: 0.7595


Após a aplicação de três algoritmos distintos de aprendizado de máquina, os resultados das acurácias para cada um foram:


Regressão Logística: 0.7342

Árvore de Decisão: 0.6203

Support Vector Machine (SVM): 0.7595

Dentre os algoritmos testados, o Support Vector Machine (SVM) apresentou a melhor acurácia, com um valor de 0.7595. Esse resultado indica que, para o conjunto de dados e pré-processamentos específicos utilizados, o SVM foi capaz de construir um modelo que distingue melhor entre as classes alvo em comparação com os outros algoritmos.

SVM é considerado o algoritmo mais adequado para este conjunto de dados específico, proporcionando uma melhor capacidade de previsão em comparação com a Regressão Logística e a Árvore de Decisão sob as mesmas condições de pré-processamento e avaliação.